packages/ec2-metadata-service/src/MetadataService.ts

```diff
@@ -0,0 +1,115 @@
+import { HttpRequest } from "@aws-sdk/protocol-http";
+import { HttpHandlerOptions } from "@aws-sdk/types";
+import { loadConfig } from "@smithy/node-config-provider";
+import { NodeHttpHandler } from "@smithy/node-http-handler";
+import { sdkStreamMixin } from "@smithy/util-stream";
+
+import { ENDPOINT_SELECTORS, IMDSv1_DISABLED_SELECTORS } from "./ConfigLoaders";
+import { MetadataServiceOptions } from "./MetadataServiceOptions";
+
+/**
+ * @public
+ */
+export class MetadataService {
+  private disableFetchToken: boolean;
+  private config: Promise<MetadataServiceOptions>;
+  /**
+   * Creates a new MetadataService object with a given set of options.
+   */
+  constructor(options: MetadataServiceOptions = {}) {
+    this.config = (async () => {
+      const profile = options?.profile || process.env.AWS_PROFILE;
+      return {
+        endpoint: options.endpoint ?? (await loadConfig(ENDPOINT_SELECTORS, { profile })()),
+        httpOptions: {
+          timeout: options?.httpOptions?.timeout || 0,
+        },
+        ec2MetadataV1Disabled:
+          options?.ec2MetadataV1Disabled ?? (await loadConfig(IMDSv1_DISABLED_SELECTORS, { profile })()),
+      };
+    })();
+    this.disableFetchToken = options?.disableFetchToken || false;
+  }
+
+  async request(path: string, options: { method?: string; headers?: Record<string, string> }): Promise<string> {
+    const { endpoint, ec2MetadataV1Disabled } = await this.config;
+    const handler = new NodeHttpHandler();
+    const endpointUrl = new URL(endpoint!);
+    const headers = options.headers || {};
+    /**
+     * If IMDSv1 is disabled and disableFetchToken is true, throw an error
+     * Refer: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html
+     */
+    if (this.disableFetchToken && ec2MetadataV1Disabled) {
+      throw new Error("IMDSv1 is disabled and fetching token is disabled, cannot make the request.");
+    }
```

Comment:
nit: newlines for readability
```suggestion

    /**
     * If IMDSv1 is disabled and disableFetchToken is true, throw an error
     * Refer: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html
     */
    if (this.disableFetchToken && ec2MetadataV1Disabled) {
      throw new Error("IMDSv1 is disabled and fetching token is disabled, cannot make the request.");
    }

```